### DISCUSSION TITLE
how to get set cookie without domain

### DISCUSSION THREAD

**Discussion by Yiming828** [curl/curl#7555] [Q&A] [answered]
Hi, I am developer from China
nice to meet you. I have one problem
when using curl, how to get set cookie if there is no domain when authenticating(log in)
thanks

**Comment by bagder** [comment, marked as answer]
Are you talking about using the command line tool or using the libcurl API?

**Reply by Yiming828** [reply]
Hi, I am developing log in automatically using C
after log in, I have to get set-cookies, but, if there is not domain, I cant catch cookies.
I am taking about libcurl API

**Reply by danielgustafsson** [reply]
If you don't have a domain, which cookie did you envision getting?

**Reply by Yiming828** [reply]
Hi, thanks for your reply. I am going to catch set-cookies(the names are _rid, _aid, XSRF-TOKEN)

**Reply by Yiming828** [reply]
If you want, I can show you as image.
![cookie_without_domain](https://user-images.githubusercontent.com/57879495/129012621-96c27f2f-2f55-4f3f-b80b-eadb52a5ce91.PNG)

**Reply by Yiming828** [reply]
right now, If there is domain, then my code works well.

**Reply by bagder** [reply]
If you tell libcurl to handle cookies in the handle, like with `CURLOPT_COOKIEFILE`, it will handle sending cookies automatically and then it'll handle cookies with or without domain set.

**Reply by Yiming828** [reply]
hi, thx for your reply. 
` req->method = method;
    strcpy(req->url, url);

    curl_easy_setopt(req->curl, CURLOPT_URL, req->url);
    curl_easy_setopt(req->curl, CURLOPT_WRITEFUNCTION, write_cb);
    curl_easy_setopt(req->curl, CURLOPT_WRITEDATA, req);
    curl_easy_setopt(req->curl, CURLOPT_COOKIEFILE, "");
` I have tried .but...

**Reply by bagder** [reply]
For that to work, you first need to **get** a resource that sets the cookies correctly, so that the subsequent request **sends** the correct cookies. Is that how you do it?

**Reply by Yiming828** [reply]
Hi, resource? excuse me, could you explain in more detail? or , can you send me sample code or link?thanks. regards.

**Reply by bagder** [reply]
Better up! I've documented how to work with cookies in libcurl: https://everything.curl.dev/libcurl-http/cookies

**Reply by Yiming828** [reply]
OK, let me check. thanks for your kind response. I have clicked "forks" and "star" for you. 
If I need help more, I will leave message here again. thanks. have good day.

**Reply by Yiming828** [reply]
HI, I have checked. in my mind, that works for given domain. so. what is solution? could you explain in more detail? thanks

**Reply by bagder** [reply]
I don't understand. A cookie as sent from a server can have the domain set or not. If not, the existing host name is used. curl works with either.

**Reply by Yiming828** [reply]
Hi, How r u doing today? could you let me know command format for no domain cookie?

**Reply by Yiming828** [reply]
I mean cod format for cookie if there is no domain. THX

**Reply by Yiming828** [reply]
*code"

**Reply by Yiming828** [reply]
yeah, I used cookielist instead of cookiefile. sign in OK. thx

**Reply by Yiming828** [reply]
sometimes success, and then failed again. what's up?

**Reply by Yiming828** [reply]
HI, good weekend? so, any solution?

**Reply by bagder** [reply]
I don't understand. A cookie as sent from a server can have the domain set or not. If not, the existing host name is used. curl works with either.

**Reply by Yiming828** [reply]
Hi, thx for your response. so, with libcurl, we can get set-cookie even though there is no domain name?

**Reply by bagder** [reply]
`set-cookie:` is a response-header that sets (saves) the cookie in the client. That header cam contain a domain name or not. Both are fine and curl works with both. As that's how cookies work.

**Reply by Yiming828** [reply]
hi thx for your reply again. how should I understand this:
"Since 7.43.0 cookies that were imported in the Set-Cookie format without a domain name are not exported by this option"

**Reply by bagder** [reply]
That's for reading cookies from a **file** stored as HTTP headers, as then we don't know for what host name/domain the file concerns. That's a discouraged way to load cookies anyway.

**Reply by bagder** [reply]
If you want to store cookies in a file, use `CURLOPT_COOKIEJAR` to make curl store them in a better format that is easier and better used when read back.

**Reply by Yiming828** [reply]
yeah, thx. but if there is no domain, file is not created when using cookiejar.

**Reply by Yiming828** [reply]
I mean cookie information is not saved in file.

**Reply by bagder** [reply]
Yes it is. Not saving the file would be a bug.

**Reply by Yiming828** [reply]
if there is no domain, then we cant get set-cookie using libcurl cookie engine now?

**Reply by bagder** [reply]
How about this:

**describe exactly what you do** and then we can discuss if that is right and what libcurl does.

If there's a problem, we need to be able to understand and reproduce. You repeating the same short sentence over and over is not driving this forward.

**Reply by Yiming828** [reply]
OK, if I use headercallback function, I can get response header. in that response header, there is set-cookie information.
1: how can I extract only set-cookie informaiton from response header?
2: after that, how can I use those extracted cookie information for other post or get requests?

**Reply by Yiming828** [reply]
[set_cookie_withoutdomain.zip](https://github.com/curl/curl/files/6990930/set_cookie_withoutdomain.zip)
please check this video

**Reply by bagder** [reply]
**You're not showing me how you use libcurl**

For the first transfer that gets the initial cookies:

1. Create an easy handle to do a libcurl transfer
2. Set [CURLOPT_COOKIEFILE](https://curl.se/libcurl/c/CURLOPT_COOKIEFILE.html) to an empty file
3. Get the initial URL
4. When the response comes back with a number of `Set-Cookie:` headers, curl will parse and store them. It knows the cookie format. **It doesn't matter if there's a specified domain or not.**: curl will handle it.

**Re-use the same handle** to do the second request that should use the cookie(s) again.
1. Setup the new URL
2. Do the transfer
3. If the cookies you received in the initial request match the new URL you're operating against, it will include those cookies in the request headers.

If you want to make things more persistent, you can make the first request use [CURLOPT_COOKIEJAR](https://curl.se/libcurl/c/CURLOPT_COOKIEJAR.html) instead. That will make curl store the cookies in that file (when the handle is closed) in addition to keeping them in memory. Then you can close the handle after the first request and in the subsequent that should use the cookies, you use [CURLOPT_COOKIEFILE](https://curl.se/libcurl/c/CURLOPT_COOKIEFILE.html) to get cookies from that previously saved cookie jar.

**Reply by Yiming828** [reply]
`http_req_t *http_create_req(http_method_t method, char *url) {
    http_req_t *req;

    req = (http_req_t *)malloc(sizeof(http_req_t));
    if (!req)
        return NULL;
    
    memset(req, 0, sizeof(*req));
    req->curl = curl_easy_init();
    if (!req->curl) {
        free(req);
        return NULL;
    }

    req->method = method;
    strcpy(req->url, url);

    curl_easy_setopt(req->curl, CURLOPT_URL, req->url);
    curl_easy_setopt(req->curl, CURLOPT_WRITEFUNCTION, write_cb);
    curl_easy_setopt(req->curl, CURLOPT_WRITEDATA, req);
    curl_easy_setopt(req->curl, CURLOPT_COOKIEFILE, "");`

**Reply by Yiming828** [reply]
that is the function for libcurl

**Reply by bagder** [reply]
Do you think it implements what I described above?

**Reply by Yiming828** [reply]
Hi, if I miss some, let me know.thx

**Reply by bagder** [reply]
You missed the whole second part that uses the cookies the first request got. It starts with **Re-use the same handle**

**Reply by bagder** [reply]
You've reached the maximum limit of responses I can do on a single question without going insane. I give up. I can continue if we agree on a paid support contract. Others are free to respond if they want. All this is documented in detail already. Good luck!

**Reply by Yiming828** [reply]
yeah, bcause I did not get cookies, I could not go the second part
